beacon_node/beacon_chain/src/sync_committee_verification.rs

```diff
@@ -0,0 +1,749 @@
+                slot: contribution.slot,
+            });
+        }
+        Ok(VerifiedSyncContribution { signed_aggregate })
+    }
+
+    /// A helper function to add this aggregate to `beacon_chain.op_pool`.
+    pub fn add_to_pool(self, chain: &BeaconChain<T>) -> Result<Self, Error> {
+        chain.add_contribution_to_block_inclusion_pool(self)
+    }
+
+    /// Returns the underlying `contribution` for the `signed_aggregate`.
+    pub fn contribution(&self) -> &SyncCommitteeContribution<T::EthSpec> {
+        &self.signed_aggregate.message.contribution
+    }
+
+    /// Returns the underlying `signed_aggregate`.
+    pub fn aggregate(&self) -> &SignedContributionAndProof<T::EthSpec> {
+        &self.signed_aggregate
+    }
+}
+
+impl VerifiedSyncSignature {
+    /// Returns `Ok(Self)` if the `sync_signature` is valid to be (re)published on the gossip
+    /// network.
+    ///
+    /// `subnet_id` is the subnet from which we received this attestation. This function will
+    /// verify that it was received on the correct subnet.
+    pub fn verify<T: BeaconChainTypes>(
+        sync_signature: SyncCommitteeSignature,
+        subnet_id: Option<SyncSubnetId>,
+        chain: &BeaconChain<T>,
+    ) -> Result<Self, Error> {
+        // Ensure sync committee signature is for the current slot (within a
+        // MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance).
+        //
+        // We do not queue future attestations for later processing.
+        verify_propagation_slot_range(chain, &sync_signature)?;
+
+        // Attestations must be for a known block. If the block is unknown, we simply drop the
+        // attestation and do not delay consideration for later.
+        //
+        // Enforce a maximum skip distance for unaggregated attestations.
+        verify_head_block_is_known(
+            chain,
+            &sync_signature,
+            sync_signature.beacon_block_root,
+            chain.config.import_max_skip_slots,
+        )?;
+        let sync_subcommittee_size =
+            <<T as BeaconChainTypes>::EthSpec as EthSpec>::SyncCommitteeSize::to_usize()
+                .safe_div(SYNC_COMMITTEE_SUBNET_COUNT as usize)?;
+        let pubkey = chain
+            .validator_pubkey_bytes(sync_signature.validator_index as usize)?
+            .ok_or(Error::UnknowValidatorIndex(
+                sync_signature.validator_index as usize,
+            ))?;
+
+        let current_sync_committee = chain.head_current_sync_committee()?;
+        let mut subnet_positions = HashMap::new();
+        for (committee_index, validator_pubkey) in current_sync_committee.pubkeys.iter().enumerate()
+        {
+            if pubkey == *validator_pubkey {
+                let subcommittee_index = committee_index.safe_div(sync_subcommittee_size)?;
+                subnet_positions
+                    .entry(SyncSubnetId::new(subcommittee_index as u64))
+                    .or_insert_with(Vec::new)
+                    .push(committee_index);
```

Comment:
I _think_ the position we should push here should be the position of the validator within the subcomittee rather than its position within the committee as a whole.

When I tried posting some sync committee signatures I generated with the VC I got errors like:

```
ERRO Unable to add sync committee signature to pool, validator_index: 63, slot: 10, error: ContributionError(SszTypesError(OutOfBounds { i: 421, len: 128 }))
```

The patch I applied to fix it was:

```rust
if pubkey == *validator_pubkey {
    let subcommittee_index = committee_index.safe_div(sync_subcommittee_size)?;
    let position_in_subcommittee = committee_index.safe_rem(sync_subcommittee_size)?;
    subnet_positions
        .entry(SyncSubnetId::new(subcommittee_index as u64))
        .or_insert_with(Vec::new)
        .push(position_in_subcommittee);
}
```

What do you reckon?

Reply:
Yep, definitely a bug! Thank you kindly!